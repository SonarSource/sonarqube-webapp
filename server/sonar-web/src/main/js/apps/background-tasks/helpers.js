/*
 * SonarQube
 * Copyright (C) 2009-2016 SonarSource SA
 * mailto:contact AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
const ONE_SECOND = 1000;
const ONE_MINUTE = 60 * ONE_SECOND;

function format (int, suffix) {
  return `${int}${suffix}`;
}

export function formatDuration (value) {
  if (!value) {
    return '';
  }
  if (value >= ONE_MINUTE) {
    const minutes = Math.round(value / ONE_MINUTE);
    return format(minutes, 'min');
  } else if (value >= ONE_SECOND) {
    const seconds = Math.round(value / ONE_SECOND);
    return format(seconds, 's');
  } else {
    return format(value, 'ms');
  }
}
