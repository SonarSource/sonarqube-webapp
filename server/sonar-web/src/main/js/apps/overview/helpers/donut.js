import d3 from 'd3';
import React from 'react';

let Sector = React.createClass({
  render() {
    let arc = d3.svg.arc()
        .outerRadius(this.props.radius)
        .innerRadius(this.props.radius - this.props.thickness);
    return <path d={arc(this.props.data)} style={{ fill: this.props.fill }}/>;
  }
});

export default React.createClass({
  getDefaultProps() {
    return {
      size: 30,
      thickness: 6
    };
  },

  render() {
    let radius = this.props.size / 2;
    let pie = d3.layout.pie()
        .sort(null)
        .value(d => d.value);
    let data = this.props.data;
    let sectors = pie(data).map((d, i) => {
      return <Sector key={i} data={d} fill={data[i].fill} radius={radius} thickness={this.props.thickness}/>;
    });
    return <svg width={this.props.size} height={this.props.size}>
      <g transform={`translate(${radius}, ${radius})`}>{sectors}</g>
    </svg>;
  }
});
