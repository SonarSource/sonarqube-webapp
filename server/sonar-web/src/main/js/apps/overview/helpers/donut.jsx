import React from 'react';

const Sector = React.createClass({
  render() {
    const arc = d3.svg.arc()
        .outerRadius(this.props.radius)
        .innerRadius(this.props.radius - this.props.thickness);
    return <path d={arc(this.props.data)} style={{ fill: this.props.fill }}></path>;
  }
});

export default React.createClass({
  getDefaultProps() {
    return {
      size: 30,
      thickness: 6
    };
  },

  render() {
    const radius = this.props.size / 2;
    const pie = d3.layout.pie().sort(null)
        .value(d => {
          return d.value
        });
    const data = this.props.data;
    const sectors = pie(data).map((d, i) => {
      return <Sector
          key={i}
          data={d}
          fill={data[i].fill}
          radius={radius}
          thickness={this.props.thickness}/>;
    });
    return (
        <svg width={this.props.size} height={this.props.size}>
          <g transform={`translate(${radius}, ${radius})`}>{sectors}</g>
        </svg>
    );
  }
});
