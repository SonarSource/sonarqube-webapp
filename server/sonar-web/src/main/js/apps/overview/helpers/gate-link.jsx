import React from 'react';

export default React.createClass({
  render() {
    const url = `${baseUrl}/quality_gates/show/${this.props.gate}`;
    return <a href={url}>{this.props.children}</a>;
  }
});
