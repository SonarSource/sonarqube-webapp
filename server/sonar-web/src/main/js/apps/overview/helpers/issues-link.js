import React from 'react';

export default React.createClass({
  render() {
    let params = Object.keys(this.props.params).map((key) => {
          return `${key}=${encodeURIComponent(this.props.params[key])}`;
        }).join('|'),
        url = `${baseUrl}/component_issues/index?id=${encodeURIComponent(this.props.component)}#${params}`;
    return <a href={url}>{this.props.children}</a>;
  }
});
