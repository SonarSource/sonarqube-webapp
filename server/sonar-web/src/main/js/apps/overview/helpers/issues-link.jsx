import React from 'react';

export default React.createClass({
  render: function () {
    var params = Object.keys(this.props.params).map((key) => {
      return `${key}=${encodeURIComponent(this.props.params[key])}`;
    }).join('|');
    var url = `${baseUrl}/component_issues/index?id=${encodeURIComponent(this.props.component)}#${params}`;
    return <a href={url}>{this.props.children}</a>;
  }
});
