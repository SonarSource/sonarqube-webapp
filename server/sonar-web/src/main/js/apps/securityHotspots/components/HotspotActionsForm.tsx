/*
 * SonarQube
 * Copyright (C) 2009-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import { assignSecurityHotspot, setSecurityHotspotStatus } from '../../../api/security-hotspots';
import {
  HotspotResolution,
  HotspotSetStatusRequest,
  HotspotStatus,
  HotspotStatusOptions,
  HotspotUpdateFields
} from '../../../types/security-hotspots';
import HotspotActionsFormRenderer from './HotspotActionsFormRenderer';

interface Props {
  hotspotKey: string;
  onSubmit: (data: HotspotUpdateFields) => void;
}

interface State {
  comment: string;
  selectedUser?: T.UserActive;
  selectedOption: HotspotStatusOptions;
  submitting: boolean;
}

export default class HotspotActionsForm extends React.Component<Props, State> {
  state: State = {
    comment: '',
    selectedOption: HotspotStatusOptions.FIXED,
    submitting: false
  };

  handleSelectOption = (selectedOption: HotspotStatusOptions) => {
    this.setState({ selectedOption });
  };

  handleAssign = (selectedUser: T.UserActive) => {
    this.setState({ selectedUser });
  };

  handleCommentChange = (comment: string) => {
    this.setState({ comment });
  };

  handleSubmit = (event: React.SyntheticEvent<HTMLFormElement>) => {
    event.preventDefault();

    const { hotspotKey } = this.props;
    const { comment, selectedOption, selectedUser } = this.state;

    const status =
      selectedOption === HotspotStatusOptions.ADDITIONAL_REVIEW
        ? HotspotStatus.TO_REVIEW
        : HotspotStatus.REVIEWED;
    const data: HotspotSetStatusRequest = { status };

    // If reassigning, ignore comment for status update. It will be sent with the reassignment below
    if (comment && !(selectedOption === HotspotStatusOptions.ADDITIONAL_REVIEW && selectedUser)) {
      data.comment = comment;
    }

    if (selectedOption !== HotspotStatusOptions.ADDITIONAL_REVIEW) {
      data.resolution = HotspotResolution[selectedOption];
    }

    this.setState({ submitting: true });
    return setSecurityHotspotStatus(hotspotKey, data)
      .then(() => {
        if (selectedOption === HotspotStatusOptions.ADDITIONAL_REVIEW && selectedUser) {
          return this.assignHotspot(selectedUser, comment);
        }
        return null;
      })
      .then(() => {
        this.props.onSubmit({ status, resolution: data.resolution });
      })
      .catch(() => {
        this.setState({ submitting: false });
      });
  };

  assignHotspot = (assignee: T.UserActive, comment: string) => {
    const { hotspotKey } = this.props;

    return assignSecurityHotspot(hotspotKey, {
      assignee: assignee.login,
      comment
    });
  };

  render() {
    const { hotspotKey } = this.props;
    const { comment, selectedOption, selectedUser, submitting } = this.state;

    return (
      <HotspotActionsFormRenderer
        comment={comment}
        hotspotKey={hotspotKey}
        onAssign={this.handleAssign}
        onChangeComment={this.handleCommentChange}
        onSelectOption={this.handleSelectOption}
        onSubmit={this.handleSubmit}
        selectedOption={selectedOption}
        selectedUser={selectedUser}
        submitting={submitting}
      />
    );
  }
}
