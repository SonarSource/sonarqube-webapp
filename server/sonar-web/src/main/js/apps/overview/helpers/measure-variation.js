import React from 'react';
import { formatMeasureVariation } from '../../../helpers/measures';


export default React.createClass({
  render() {
    if (this.props.value == null || isNaN(this.props.value)) {
      return null;
    }
    let formatted = formatMeasureVariation(this.props.value, this.props.type);
    return <span>{formatted}</span>;
  }
});
