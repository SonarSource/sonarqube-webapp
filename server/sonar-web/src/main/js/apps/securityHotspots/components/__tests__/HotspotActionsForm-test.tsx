/*
 * SonarQube
 * Copyright (C) 2009-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { shallow } from 'enzyme';
import * as React from 'react';
import { waitAndUpdate } from 'sonar-ui-common/helpers/testUtils';
import { assignSecurityHotspot, setSecurityHotspotStatus } from '../../../../api/security-hotspots';
import { mockLoggedInUser } from '../../../../helpers/testMocks';
import {
  HotspotResolution,
  HotspotStatus,
  HotspotStatusOption
} from '../../../../types/security-hotspots';
import HotspotActionsForm from '../HotspotActionsForm';

jest.mock('../../../../api/security-hotspots', () => ({
  assignSecurityHotspot: jest.fn().mockResolvedValue(undefined),
  setSecurityHotspotStatus: jest.fn().mockResolvedValue(undefined)
}));

it('should render correctly', () => {
  expect(shallowRender()).toMatchSnapshot();
});

it('should handle option selection', () => {
  const wrapper = shallowRender();
  expect(wrapper.state().selectedOption).toBe(HotspotStatusOption.FIXED);
  wrapper.instance().handleSelectOption(HotspotStatusOption.SAFE);
  expect(wrapper.state().selectedOption).toBe(HotspotStatusOption.SAFE);
});

it('should handle comment change', () => {
  const wrapper = shallowRender();
  wrapper.instance().handleCommentChange('new comment');
  expect(wrapper.state().comment).toBe('new comment');
});

it('should handle submit', async () => {
  const onSubmit = jest.fn();
  const wrapper = shallowRender({ onSubmit });
  wrapper.setState({ selectedOption: HotspotStatusOption.ADDITIONAL_REVIEW });
  await waitAndUpdate(wrapper);

  const preventDefault = jest.fn();
  const promise = wrapper.instance().handleSubmit({ preventDefault } as any);
  expect(preventDefault).toBeCalled();

  expect(wrapper.state().submitting).toBe(true);
  await promise;
  expect(setSecurityHotspotStatus).toBeCalledWith('key', {
    status: HotspotStatus.TO_REVIEW
  });
  expect(onSubmit).toBeCalled();

  // SAFE
  wrapper.setState({ comment: 'commentsafe', selectedOption: HotspotStatusOption.SAFE });
  await waitAndUpdate(wrapper);
  await wrapper.instance().handleSubmit({ preventDefault } as any);
  expect(setSecurityHotspotStatus).toBeCalledWith('key', {
    comment: 'commentsafe',
    status: HotspotStatus.REVIEWED,
    resolution: HotspotResolution.SAFE
  });

  // FIXED
  wrapper.setState({ comment: 'commentFixed', selectedOption: HotspotStatusOption.FIXED });
  await waitAndUpdate(wrapper);
  await wrapper.instance().handleSubmit({ preventDefault } as any);
  expect(setSecurityHotspotStatus).toBeCalledWith('key', {
    comment: 'commentFixed',
    status: HotspotStatus.REVIEWED,
    resolution: HotspotResolution.FIXED
  });
});

it('should handle assignment', async () => {
  const onSubmit = jest.fn();
  const wrapper = shallowRender({ onSubmit });
  wrapper.setState({
    comment: 'assignment comment',
    selectedOption: HotspotStatusOption.ADDITIONAL_REVIEW
  });

  wrapper.instance().handleAssign(mockLoggedInUser({ login: 'userLogin' }));
  await waitAndUpdate(wrapper);

  const promise = wrapper.instance().handleSubmit({ preventDefault: jest.fn() } as any);

  expect(wrapper.state().submitting).toBe(true);
  await promise;

  expect(setSecurityHotspotStatus).toBeCalledWith('key', {
    status: HotspotStatus.TO_REVIEW
  });
  expect(assignSecurityHotspot).toBeCalledWith('key', {
    assignee: 'userLogin',
    comment: 'assignment comment'
  });
  expect(onSubmit).toBeCalled();
});

it('should handle submit failure', async () => {
  const onSubmit = jest.fn();
  (setSecurityHotspotStatus as jest.Mock).mockRejectedValueOnce('failure');
  const wrapper = shallowRender({ onSubmit });
  const promise = wrapper.instance().handleSubmit({ preventDefault: jest.fn() } as any);
  expect(wrapper.state().submitting).toBe(true);
  await promise.catch(() => {});
  expect(wrapper.state().submitting).toBe(false);
  expect(onSubmit).not.toBeCalled();
});

function shallowRender(props: Partial<HotspotActionsForm['props']> = {}) {
  return shallow<HotspotActionsForm>(
    <HotspotActionsForm hotspotKey="key" onSubmit={jest.fn()} {...props} />
  );
}
