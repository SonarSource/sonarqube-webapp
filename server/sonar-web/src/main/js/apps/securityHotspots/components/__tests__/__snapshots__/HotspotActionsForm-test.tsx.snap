// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<HotspotActionsFormRenderer
  hotspotKey="key"
  onAssign={[Function]}
  onSelectOption={[Function]}
  onSubmit={[Function]}
  selectedOption="FIXED"
  submitting={false}
/>
`;
